### ISSUE TITLE
local tests failing with "too many open files"

### ISSUE THREAD

**Issue by rabernat** [xgcm/xmitgcm#156]
I can no longer run the test suite locally. Tests start failing with "OSError: [Errno 24] Too many open files". @aulemahal found the same problem in #149.

We must not be closing files somewhere.

**Comment by malmans2** [comment]
I have the same problem on my machines. It looks like the failure only occurs when both `np.memmap` and `dask` are used, so I think `dask` is not closing the `memmap` references. 
Reading [this comment](https://github.com/astropy/ccdproc/issues/629#issuecomment-399507854) it looks like `memmap` is creating three references, and [this comment](https://stackoverflow.com/questions/5830397/python-mmap-error-too-many-open-files-whats-wrong) suggests to use `.copy` to make sure that the references are closed.

I can fix all tests using `.copy()` in the `load_chunk()` functions when `use_dask is True` and `use_mmap is True`. Here is one of the `load_chunk()` functions defined: https://github.com/xgcm/xmitgcm/blob/94c7f8ac013a56f3ba4991dc353955ee7001d7c9/xmitgcm/utils.py#L549-L551
I'm not sure if that's the right way to go though, and if I'm just getting rid of the advantages of using `dask` and `mmap`.

**Comment by rabernat** [comment]
Thanks for digging into this.

One option would be to make dask and memmap mutually exclusive. There is actually no good reason to use memmap with dask, because whenver dask loads a chunk, it loads it as numpy array. Memmap is in effect a different way of being "lazy" with the data loading.

> I'm not sure if that's the right way to go though

I think your example code just leads to eager loading, so at that point you might as well just say `memmap=False`.

Instead I would try changing the logic in `utils.read_mds`:
https://github.com/xgcm/xmitgcm/blob/94c7f8ac013a56f3ba4991dc353955ee7001d7c9/xmitgcm/utils.py#L87-L89

Currently, both `use_dask` and `use_mmap` default to `True. We could change the default to `use_mmap=None` and then have some logic in the function to disable `use_mmap` when `use_dask` is set to True.

I'm curious to know if this would resolve the "too many open files" test failures.

**Comment by malmans2** [comment]
Yes! This fixes the problem:
```python
    # Do not use mmap with dask   
    if use_mmap is None:
        use_mmap = False if (use_dask) else True
```
Do you also want to set `use_mmap=False` (and maybe add a warning) if `use_mmap=True` and `use_dask=True`?

**Comment by rabernat** [comment]
I think we want the following sort of behavior for the keyword args to `read_mds`

| use_dask | use_memmap | result |
| --- | --- | --- |
| True | False | default behavior |
| True | True | Raise an error explaining it's pointless |
| True | None | use_memmap get sets to False |
| False | False | Eager loading to numpy arrays |
| False | True | Load data to memmap |
| False | None | use_memmap get set to True |

I think this is reasonably compatible with the current behavior. The default values for `read_mds` should be `use_dask=True, use_mmap=None`.

**Comment by aulemahal** [comment]
I don't know if this should be another issue or not, but I've got a similar problem now.
With #165 added to my local branch, I still get `no space left on device` and my computer crashes when testing twice.

Turns out my `/tmp` folder is too small. Running py.test once fills it with 3GB of data. 
A workaround would be me setting a different "basetemp" when calling py.test, or me simply `rm -rf` everything after running it.

Or, we could do something similar to : https://stackoverflow.com/a/51593679/9291575

Is the current behaviour the correct one?
